conn/physic/units.go

```diff
@@ -733,3 +734,250 @@ func picoAsString(v int64) string {
 	}
 	return sign + strconv.Itoa(base) + "." + prefixZeros(3, frac) + unit
 }
+
+// Decimal is the exact representation of decimal number.
+type decimal struct {
+	// digits hold the string representation of the significant decimal digits.
+	digits string
+	// exponent is the left or right decimal shift. (powers of ten).
+	exp int
+	// neg it true if the number is negative.
+	neg bool
+}
+
+// Positive powers of 10 in the form such that powerOF10[index] = 10^index.
+var powerOf10 = [19]uint64{
+	1,
+	10,
+	100,
+	1000,
+	10000,
+	100000,
+	1000000,
+	10000000,
+	100000000,
+	1000000000,
+	10000000000,
+	100000000000,
+	1000000000000,
+	10000000000000,
+	100000000000000,
+	1000000000000000,
+	10000000000000000,
+	100000000000000000,
+	1000000000000000000,
+}
+
+// Converts from decimal to int64, using the decimal.digits character values and
+// converting to a intermediate unit64.
+// Scale is combined with the decimal exponent to maximise the resolution and is
+// in powers of ten.
+func dtoi(d decimal, scale int) (int64, error) {
+	// Maximum value for a int64.
+	const max = (1<<63 - 1)
+	// Use uint till the last as it allows checks for overflows.
+	var u uint64
+	for i := 0; i < len(d.digits); i++ {
+		// Check that is is a digit.
+		if d.digits[i] >= '0' && d.digits[i] <= '9' {
+			// '0' = 0x30 '1' = 0x31 ...etc.
+			digit := d.digits[i] - '0'
+			// *10 is decimal shift left.
+			u *= 10
+			check := u + uint64(digit)
+			// Check should always be larger than u unless we have overflowed.
+			// Similarly if check > max it will overflow when converted to int64.
+			if check < u || check > max {
+				return int64(u), &parseError{err: errors.New("overflows maximum is"), s: strconv.FormatUint(max, 10)}
```

Comment:
I'd prefer if the FormatUint(max) was a constant. Maybe a global. So basically an option is:

```
const max = (1<<63-1)
var maxStr = "9223372036854775807" 
```

then in test do
```
func TestingMax(t* testing.T) {
  if strconv.FormatUint(max, 10) != maxStr {
    t.Fatal("unexpected text representation of max")
  }
}
```

I know this is optimizing breadcrumbs but I really try to reduce the number of runtime calls.

Reply:
no I like it, cleaner too. I made it maxInt64 and maxInt64str for a bit of extra clarity since it is global.